src/compiler/crystal/syntax/lexer.cr

```diff
@@ -127,64 +127,35 @@ module Crystal
             next_char :"<<="
           when '-'
             here = StringIO.new(20)
-            here_start = 0
 
             while true
               case char = next_char
               when '\n'
                 @line_number += 1
                 @column_number = 0
-                next_char
-                here_start = current_pos
                 break
+              when '\\'
+                if peek_next_char == 'n'
+                  next_char
+                  raise "invalid heredoc identifier"
+                end
+              when ' '
+                case peek_next_char
+                when ' '
+                  next_char
+                when '\n'
+                  next_char
+                  break
+                else
+                  raise "invalid heredoc identifier"
+                end
```

Comment:
Thanks for the feedback! Here's my reasoning to your questions:

1. Previously adding an explicit `\n` ([see here](http://play.crystal-lang.org/#/r/9kd)) to the end of a heredoc opening crashed the compiler. The code just detects if an explicit newline directly follows the opening and raises an error if so.

2. A heredoc opening should still be valid when directly followed by space characters, then a newline (non-explicit). For example, [this code](http://play.crystal-lang.org/#/r/9no) should be valid, but it raises `unterminated heredoc` without the checks.